tests/components/calendar/test_trigger.py

```diff
@@ -0,0 +1,374 @@
+"""Tests for the calendar automation.
+
+The tests create calendar based automations, set up a fake set of calendar
+events, then advance time to exercise that the automation is called. The
+tests use a fixture that mocks out events returned by the calendar entity,
+and create events using a relative time offset and then advance the clock
+forward exercising the triggers.
+"""
+from __future__ import annotations
+
+from collections.abc import Callable
+import datetime
+import logging
+import secrets
+from typing import Any, Generator
+from unittest.mock import patch
+
+import pytest
+
+from homeassistant.components import calendar
+import homeassistant.components.automation as automation
+from homeassistant.components.calendar.trigger import EVENT_START
+from homeassistant.core import HomeAssistant
+from homeassistant.setup import async_setup_component
+import homeassistant.util.dt as dt_util
+
+from tests.common import async_fire_time_changed, async_mock_service
+
+_LOGGER = logging.getLogger(__name__)
+
+
+CALENDAR_ENTITY_ID = "calendar.calendar_2"
+CONFIG = {calendar.DOMAIN: {"platform": "demo"}}
+
+TEST_AUTOMATION_ACTION = {
+    "service": "test.automation",
+    "data_template": {
+        "platform": "{{ trigger.platform }}",
+        "event": "{{ trigger.event }}",
+        "calendar_event": "{{ trigger.calendar_event }}",
+    },
+}
+
+# The trigger sets two alarms: One based on the next event and one
+# to refresh the schedule. The test advances the time an arbitrary
+# amount to trigger either type of event with a small jitter.
+TEST_TIME_ADVANCE_INTERVAL = datetime.timedelta(minutes=1)
+TEST_UPDATE_INTERVAL = datetime.timedelta(minutes=15)
+
+
+@pytest.fixture
+async def now() -> datetime.datetime:
+    """Fixture to provide a base time for tests."""
+    return dt_util.utcnow()
+
+
+class FakeSchedule:
+    """Test fixture class for return events in a specific date range."""
+
+    def __init__(self, now: datetime.datetime):
+        """Initiailize FakeSchedule."""
+        # Map of event start time to event
+        self.events: list[dict[str, Any]] = []
+        self.now = now
+
+    def create_event(
+        self, start_timedelta: datetime.timedelta, end_timedelta: datetime.timedelta
+    ) -> dict[str, Any]:
+        """Create a new fake event, used by tests."""
+        event_data = {
+            "title": "Event %s" % secrets.token_hex(16),  # Arbitrary unique data
```

Comment:
Please use f-strings.

Reply:
Is the rationale: Use f-strings whenever the expression is reasonable,  except when it might cause extra performance overhead when not rendered. During tests: always running so use f-strings? I've left debug statements within the code alone for example, but converted the debug statements in the tests to f-strings.

Reply:
We only use old string format operator for logging statements. Sometimes we use `str.format` in constants. Otherwise we use f-strings.